common/util/auto_pop_stack.h

```diff
@@ -0,0 +1,107 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+#define VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+
+#include <iterator>
+#include <type_traits>
+#include <vector>
+#include "common/util/iterator_adaptors.h"
+
+namespace verible {
+
+// AutoPop stack class. Templated version of SyntaxTreeContext::AutoPop
+template <typename T>
+class AutoPopStack {
+ public:
+  typedef T value_type;
+  typedef typename std::remove_pointer<value_type>::type basic_value_type;
+  typedef AutoPopStack<value_type> this_type;
+
+  // member class to handle push and pop of stack safely
+  class AutoPop {
+   public:
+    AutoPop(this_type* stack, value_type&& value) {
+      stack_ = stack;
+      stack->Push(std::move(value));
+    }
+    AutoPop(this_type* stack, basic_value_type& value) {
+      stack_ = stack;
+      stack->Push(&value);
+    }
+    ~AutoPop() {
+      stack_->Pop();
+    }
+
+   private:
+    this_type* stack_;
+  };
+
+  // returns depth of context stack
+  size_t size() const { return stack_.size(); }
+
+  // returns true if the stack is empty
+  bool empty() const { return stack_.empty(); }
+
+  // returns the top value_type of the stack
+  const value_type& top() const {
+    CHECK(!stack_.empty());
+    return stack_.back();
+  }
+
+  // All elements of stack are non-null because they refer to traversed nodes.
+  typedef std::vector<value_type> stack_type;
+  typedef typename stack_type::iterator iterator;
+  typedef typename stack_type::const_iterator const_iterator;
+  typedef typename stack_type::const_reverse_iterator const_reverse_iterator;
+
+  // Allow read-only random-access into stack:
+  const_iterator begin() const { return stack_.begin(); }
+  const_iterator end() const { return stack_.end(); }
+
+  // Reverse iterators be useful for searching from the top-of-stack downward.
+  const_reverse_iterator rbegin() const { return stack_.rbegin(); }
+  const_reverse_iterator rend() const { return stack_.rend(); }
+
+ protected:
+  value_type& top() {
+    // Efficient C++, 3rd, Item 3, "Use const whenever possible."
+    return const_cast<value_type&>(
+        static_cast<const this_type&>(*this).top());
+  }
```

Comment:
No need for casting, just `return stack_.back();`.  This should internally look identical to the `const` method counterpart.

Reply:
Just wanted to reuse `const` `top()` method. Fixed.